### PR TITLE
Add StakeInstruction::DeactivateDelinquent

### PR DESCRIPTION
**Pull request by mvines**
There's no mechanism to cleanup stake delegated to abandoned vote accounts. As a motivating example, 9% of the devnet stake has now been abandoned for almost 100 epochs. Currently the only path to remove that 9% dead weight is by a network restart.

This PR proposes a new stake program instruction that permits anybody to deactivate stake delegated to a vote account that has not landed a vote in the last 5 (<s>10</s>) epochs. There's no reward for performing this common good so there should be no motivation to bot this instruction as soon as a stake account becomes eligible for public destaking.

The benefit of this approach over some form of automatic destaking by the runtime is the lack of complexity.  The stake program changes here are very minimal and thus easy to inspect and get comfortable with (or not!).

TODO:
* [x] Decide if 10 epochs is the right magic number, which is ~three or more weeks in human time. This feels long enough for an operator to resolve any temporary censorship attempts by another validator/staker that's attempting to force them off the network. **Update**: I thought about it and 5 epochs seems like plenty of buffer, which is approximately 15 days of delinquency.
* [x] Add tests if this approach is accepted
* [x] Cli support
* [x] Vote accounts that have never voted but have managed to accumulate active stake somehow are currently excluded from this instruction. Decide if this is acceptable. The alternative would be to permit such stake to be deactivated immediately. **Update**: Stake delegated to a vote account that has never voted may also be destaked by anybody
* [x] Consider requiring a second vote account that voted in each of the last 5 (<s>10</s>) epochs be provided to the instruction. This second vote account acts as proof to the stake program that the chain has been active over the last 5 (<s>10</s>) epoch, and prevents a very unlikely vector where a stake account could be destaked if the attacker is able to warp the chain ahead by 5 (<s>10</s>) epochs (perhaps on a mainnet restart somehow). This might be overkill. **Update**: Added

### CODE REVIEW

**Comment by leoluk** [comment]
Big :+1: on this - jailing has been working very well on Cosmos networks.

> Decide if 10 epochs is the right magic number, which is ~three or more weeks in human time. This feels long enough for an operator to resolve any temporary censorship attempts by another validator/staker that's attempting to force them off the network.

What's the current threshold for a successful censorship attempt? Unless it's very low, I'd go with a much more aggressive threshold, like 2-3 epochs. It's a good way to further incentivize higher-quality validator operations.

Related issue:
- https://github.com/solana-labs/solana/issues/13458.

**Comment by mvines** [comment]
> What's the current threshold for a successful censorship attempt? Unless it's very low, we could even go a much more aggressive threshold, like 2-3 epochs - this is still plenty of time. It's a good way to further incentivize higher-quality validator operations.

I feel the threshold would be pretty high, but only have intuition. To prevent a validator from voting at least once in multiple epochs seems like it would require a large percentage of the stake to reject all vote transactions and all blocks containing vote transactions from the victim. To be fully effective, I think 2/3rds of the stake would need be participating at which point it doesn't matter anyway.

> Did we previously use the deliquency status for anything critical?

Beyond obviously computing epoch rewards, no.

**Comment by mvines** [comment]
I think this is ready for review, but no rush as I’m OOO for the next week or so.

**Review by t-nelson — COMMENTED** [review]
cursory look at the logic looks good.  i think we can improve the error reporting so clients can be more helpful with failures

File: cli/src/stake.rs [hunk 1]
```diff
@@ -1486,19 +1496,120 @@ pub fn process_deactivate_stake_account(
...
+                for (vote_epoch, ..) in epoch_credits[epoch_index..].iter().rev() {
+                    if *vote_epoch != epoch {
+                        return false;
+                    }
+                    if epoch == 0 {
+                        return false;
+                    }
+                    epoch -= 1;
+                }
+                true
```
[the remark this answers is not in the document]
Comment (mvines) [reply cli/src/stake.rs]: Added a `solana_program::stake::acceptable_reference_epoch_credits()` helper

**Comment by mvines** [comment]
> cursory look at the logic looks good. i think we can improve the error reporting so clients can be more helpful with failures

Added some additional `StakeError`s in a couple places instead of using more general `InstructionError`s

**Comment by CriesofCarrots** [comment]
Shoot, I forgot to review while you were gone. Will look a little later this afternoon!

**Comment by mvines** [comment]
CI is now happy here other than the unrelated downstream-projects timeout

File: programs/stake/src/stake_instruction.rs [hunk 2]
```diff
@@ -6128,4 +6217,265 @@ mod tests {
...
+        process_instruction_deactivate_delinquent(
+            &stake_address,
+            &stake_account,
+            &vote_account,
+            &reference_vote_account,
+            Ok(()),
+        );
+
+        // `reference_vote_account` has consistently voted and `vote_account` has voted once the
+        // last `MINIMUM_DEACTIVATE_DELINQUENT_EPOCHS`.
```
Comment (CriesofCarrots) [inline programs/stake/src/stake_instruction.rs:6465]:
```suggestion
        // `reference_vote_account` has consistently voted and `vote_account` voted once
        // `MINIMUM_DEACTIVATE_DELINQUENT_EPOCHS` - 1 epochs ago
```

File: programs/stake/src/stake_instruction.rs [hunk 3]
```diff
@@ -6128,4 +6217,265 @@ mod tests {
...
+        process_instruction_deactivate_delinquent(
+            &unrelated_stake_address,
+            &unrelated_stake_account,
+            &vote_account,
+            &reference_vote_account,
+            Err(StakeError::VoteAddressMismatch.into()),
+        );
+
+        // `reference_vote_account` has consistently voted and `vote_account` has voted once the
+        // last `MINIMUM_DEACTIVATE_DELINQUENT_EPOCHS` - 1.
```
Comment (CriesofCarrots) [inline programs/stake/src/stake_instruction.rs:6449]:
```suggestion
        // `reference_vote_account` has consistently voted and `vote_account` voted once
        // `MINIMUM_DEACTIVATE_DELINQUENT_EPOCHS` ago.
```

**Review by CriesofCarrots — COMMENTED** [review]
A couple nitty comments on the runtime side of things. Looking at cli now

File: cli/src/stake.rs [hunk 4]
```diff
@@ -4174,6 +4249,34 @@ mod tests {
...
+                    nonce_authority: 0,
+                    memo: None,
+                    seed: None,
+                    fee_payer: 0,
+                },
+                signers: vec![read_keypair_file(&default_keypair_file).unwrap().into()],
+            }
+        );
+
+        // Test DeactivateStake Subcommand with deserialized flag
```
Comment (CriesofCarrots) [inline cli/src/stake.rs:4265]:
```suggestion
        // Test DeactivateStake Subcommand with delinquent flag
```

File: cli/src/stake.rs [hunk 5]
```diff
@@ -1486,19 +1496,84 @@ pub fn process_deactivate_stake_account(
...
         Pubkey::create_with_seed(stake_account_pubkey, seed, &stake::program::id())?
     } else {
         *stake_account_pubkey
     };
 
-    let ixs = vec![stake_instruction::deactivate_stake(
-        &stake_account_address,
-        &stake_authority.pubkey(),
-    )]
+    let ixs = vec![if deactivate_delinquent {
```
Comment (CriesofCarrots) [inline cli/src/stake.rs:1506]: Gate all this business by `if !sign_only`? Might also be nice to make the `delinquent` cli arg conflict with `SIGN_ONLY_ARG.name`.

File: cli/src/stake.rs [hunk 6]
```diff
@@ -379,6 +379,12 @@ impl StakeSubCommands for App<'_, '_> {
                         .help("Seed for address generation; if specified, the resulting account \
                                will be at a derived address of STAKE_ACCOUNT_ADDRESS")
                 )
+                .arg(
+                    Arg::with_name("delinquent")
```
Comment (CriesofCarrots) [inline cli/src/stake.rs:383]: I wish we could do this without the parameter, but I guess it's too much to check the vote account for delinquency on every deactivation attempt...

File: programs/stake/src/stake_instruction.rs [hunk 7]
```diff
@@ -6128,4 +6217,265 @@ mod tests {
...
+                .deactivation_epoch,
+            current_epoch
+        );
+
+        // `reference_vote_account` has consistently voted and `vote_account` has not voted for the
+        // last `MINIMUM_DEACTIVATE_DELINQUENT_EPOCHS`. Try to deactivate an unrelated stake
+        // account.  Instruction will fail
+        let unrelated_vote_address = Pubkey::new_unique();
+        let unrelated_stake_address = Pubkey::new_unique();
+        let mut unrelated_stake_account = stake_account.clone();
```
Comment (t-nelson) [inline programs/stake/src/stake_instruction.rs:6426]:
```suggestion
        let mut unrelated_stake_account = stake_account.clone();
        assert_ne!(unrelated_vote_address, vote_address);
```

File: programs/stake/src/stake_instruction.rs [hunk 8]
```diff
@@ -6128,4 +6217,265 @@ mod tests {
...
+            .serialize_data(&VoteStateVersions::new_current(vote_state))
+            .unwrap();
+        process_instruction_deactivate_delinquent(
+            &stake_address,
+            &stake_account,
+            &vote_account,
+            &reference_vote_account,
+            Err(StakeError::MinimumDeactivateDelinquentEpochsNotMet.into()),
+        );
+    }
```
Comment (t-nelson) [inline programs/stake/src/stake_instruction.rs:6480]: test for feature inactive fails?

**Review by t-nelson — COMMENTED** [review]
looking good. great test coverage!

seems like the cli command that we really want is one that takes the delinquent and reference vote accounts, then looks up the delinquent voter's stake accounts and deactivates them all.

File: cli/src/stake.rs (hunk 6 above)
Comment (mvines) [reply cli/src/stake.rs:383]: I also wanted this to be explicit by the user as a way to ensure the responsibly for performing the action is clearly with the user

File: programs/stake/src/stake_instruction.rs (hunk 8 above)
Comment (mvines) [reply programs/stake/src/stake_instruction.rs:6480]: I thought about that, and found the excuse that we don't do this for any other gated stake instructions welcoming.

File: sdk/program/src/stake/mod.rs [hunk 9]
```diff
@@ -6,6 +6,126 @@ pub mod program {
     crate::declare_id!("Stake11111111111111111111111111111111111111");
 }
 
+use crate::clock::Epoch;
```
[the remark this answers is not in the document]
Comment (CriesofCarrots) [reply sdk/program/src/stake/mod.rs:9]: Nice, that seems much more intuitive.

**Comment by mvines** [comment]
> seems like the cli command that we really want is one that takes the delinquent and reference vote accounts, then looks up the delinquent voter's stake accounts and deactivates them all.

Yeah you're right. I figured that could be bashed though since hopefully this isn't a feature that's used regularly

File: programs/stake/src/stake_instruction.rs (hunk 2 above)
Comment (CriesofCarrots) [reply programs/stake/src/stake_instruction.rs]: I think this one got missed, prob due to the const renaming. It's at L6466-L6467 now.

Comment (mvines) [reply programs/stake/src/stake_instruction.rs]: got it this time!

**Review by CriesofCarrots — APPROVED** [review]
Looks good to me!

**Comment by mvines** [comment]
are you good here as well @t-nelson?

**Review by t-nelson — APPROVED** [review]
lgtm! :shipit:

**Pull request merged**